framework/src/main/java/org/checkerframework/common/aliasing/AliasingVisitor.java

```diff
@@ -271,7 +271,13 @@ private boolean canBeLeaked(Tree exp) {
         AnnotatedTypeMirror type = atypeFactory.getAnnotatedType(exp);
         boolean isMethodInvocation = exp.getKind() == Kind.METHOD_INVOCATION;
         boolean isNewClass = exp.getKind() == Kind.NEW_CLASS;
-        return type.hasExplicitAnnotation(Unique.class) && !isMethodInvocation && !isNewClass;
+        boolean isArray = type.getUnderlyingType().getKind().toString().equalsIgnoreCase("array");
+        boolean isNull = exp.getKind() == Kind.NULL_LITERAL;
+        if (type.getUnderlyingType().toString().startsWith("java.lang") || isArray || isNull) {
```

Comment:
What is the motivation for this special logic? You should explain the logic here or at least in the PR description.

Reply:
Also, do add tests that cover such special logic, in at least a few important cases.

Reply:
The main logic I was following that the Aliasing Checker doesn't check for explicit annotations for user defined classes. So only those classes that the user has specifically annotated as @Unique would construct @Unique classes. This would also prevent all trivial objects like strings, int, object, float, etc. from always becoming @Unique (it is more convenient to explicitly annotate them). 

However, I couldn't figure out how to make the AliasingVisitor check whether a class was user-defined or not. So, instead I listed classes and data types for which one would need explicit annotations. These included java.lang classes, arrays and null values. All this complicated conditions could be avoided if the Visitor could check for user-defined classes.

if (annotated class is user defined) {
            return type.hasAnnotation(Unique.class) && !isMethodInvocation && !isNewClass;
        } 
else {
            return type.hasExplicitAnnotation(Unique.class) && !isMethodInvocation && !isNewClass;
        }

Reply:
Can you clarify what you mean with "user defined"?
Why should the handling of explicit annotations depend on whether the class is "user defined"?

Reply:
In the stub file stubfile.astub in tests/aliasing, Object and String classes are annotated as @Unique by default. So, if I don't check for explicit annotations for all classes, then, even trivial operations like string assignments give a "unique.leaked" error in the test cases.

// part of TypeRefinementTest.java in tests/aliasing
void rule1() {
        String unique = new String();
        isUnique(unique);
        String notUnique = unique;  // gives unique.leaked error even though @Unique is never used
}

So I decided to limit this checking for "user-defined" classes (classes that are not part of any in-built java library, like java.lang or java.util). For example, in the code in the feature request #3313 (https://github.com/typetools/checker-framework/issues/3313), class Data and Demo are user-defined classes, and since Data is annotated as @Unique, all its instances don't require explicit annotations to be unique as well.

Reply:
We do not want different behavior for built-in classes and user-defined classes. The same logic should be applied to all source code, regardless of who wrote the code.

There is a difference between classes `String` and `Data` in these two examples.
In `String` the default constructor is annotated as returning a unique object. The class declaration is not annotated. Therefore, it is fine in the test case to have a non-unique String reference.
On the other hand class `Data` in this new test case is annotated as `@Unique` requiring that only `@Unique` references exist.
So your check should be for that difference between the two classes.

Also can you move the annotations from `framework/tests/aliasing/stubfile.astub` into https://github.com/typetools/jdk/pull/52?

Reply:
I have added a new method for the above requested logic. Now the code doesn't need to deal with the special cases of arrays, null and java.lang classes. I have used the getDeclAnnotations method to get the set of declared annotations of the class as a whole, and then searched for the unique annotation mirror in the set. The code now gives leaked errors without explicit annotations only when the class is annotated as @Unique. Annotating just the constructor will allow non-unique object references.